### PR TITLE
Non GET routes

### PR DESCRIPTION
**Pull request by styliii**
POST /workflows/1/process/1/steps - creates manual step
PUT /steps/1/complete - when user checks a step done, or clicks CTA to complete.
PUT /steps/1/uncomplete - when user unchecks a step
PUT /processes/1/assign - when a user assigns to a person

### CODE REVIEW

**Comment by keithtom** [comment]
Looking good!

File: gems/workflow/db/migrate/20221026011432_add_external_identifiers.rb [hunk 1]
```diff
@@ -1,12 +0,0 @@
-class AddExternalIdentifiers < ActiveRecord::Migration[7.0]
```
Comment (styliii) [inline gems/workflow/db/migrate/20221026011432_add_external_identifiers.rb:1]: this was a duplicate file.

File: spec/rails_helper.rb [hunk 2]
```diff
@@ -12,7 +12,7 @@
 require 'factory_bot_rails'
 
 FactoryBot.definition_file_paths << File.join(File.dirname(__FILE__), 'factories')
-FactoryBot.find_definitions
+# FactoryBot.find_definitions
```
Comment (styliii) [inline spec/rails_helper.rb:15]: Had to comment out this line, otherwise it kept complaining that the Address factory was already defined.

Comment (keithtom) [reply spec/rails_helper.rb:15]: weird.  let me know if you think we should just kill the gem, move models/migrations and services into core app...

File: gems/workflow/app/services/workflow/instance/step/complete.rb [hunk 3]
```diff
@@ -12,9 +12,20 @@ def initialize(step)
...
+        process = @step.process
+        prev_process_status = process.status
+
         @step.completed = true
         @step.completed_at = DateTime.now
         @step.save!
+
+        if process.status == Workflow::Instance::Process::DONE
+          process.completed_at = DateTime.now
+        end
```
Comment (keithtom) [inline gems/workflow/app/services/workflow/instance/step/complete.rb:24]: sorry not following here.  don't you have to check if the step you completed was the last step of the process? if so then mark process completed?

File: gems/workflow/app/models/workflow/instance/process.rb [hunk 4]
```diff
@@ -23,9 +27,18 @@ def effort
       super || self.definition.effort
     end
 
-    # TODO
     def status
```
Comment (styliii) [inline gems/workflow/app/models/workflow/instance/process.rb:30]: @keithtom this function will check all the steps are complete, and return the `DONE` status.

File: spec/rails_helper.rb (hunk 2 above)
Comment (styliii) [reply spec/rails_helper.rb:15]: i'm not sure if that's related to the gem, but maybe?

File: gems/workflow/app/services/workflow/instance/step/complete.rb (hunk 3 above)
Comment (styliii) [reply gems/workflow/app/services/workflow/instance/step/complete.rb:24]: the `Process#status` function will do that

File: Gemfile.lock [hunk 5]
```diff
@@ -352,7 +352,7 @@ DEPENDENCIES
   byebug
   database_cleaner-active_record
   dotenv-rails
-  elasticsearch (~> 8.x)
+  elasticsearch (< 7.14)
```
Comment (keithtom) [inline Gemfile.lock:355]: just fyi production uses 8.x but I recall having trouble installing it locally...

File: gems/workflow/app/services/workflow/instance/step/complete.rb (hunk 3 above)
Comment (keithtom) [reply gems/workflow/app/services/workflow/instance/step/complete.rb:24]: got it! thx

File: app/controllers/v1/workflow/steps_controller.rb [hunk 6]
```diff
@@ -1,8 +1,43 @@
...
+    render json: V1::Workflow::StepSerializer.new(@step)
+  end
+
+  def uncomplete
+    # TODO: identify current user, check if process/step id is accessible to user
+    @process = Workflow::Instance::Process.find_by!(external_identifier: params[:process_id])
+    @step = @process.steps.find_by!(external_identifier: params[:id])
+
+    uncompleter = Workflow::Instance::Step::Uncomplete.new(@step)
+    uncompleter.run
```
Comment (keithtom) [inline app/controllers/v1/workflow/steps_controller.rb:33]: just FYI you can run `Workflow::Instance::Step::Uncomplete.run(@step)`
The base service provides a little magic for this where `.run` does the `#new` for you

File: gems/workflow/app/services/workflow/instance/step/complete.rb (hunk 3 above)
Comment (keithtom) [reply gems/workflow/app/services/workflow/instance/step/complete.rb:24]: I think status is a useful front end concept that we should put in the serializer.  but in the back end I want to explicitly call out two separate things: your completeness and whether your dependencies are met are not.

I think of serializers as "presentation logic" where you talk in more end user terms.  and I try to keep the rest of the code more engineering-level precise.

**Review by keithtom — APPROVED** [review]
This is overall pretty good.  There's some changes we should consider but can also be part of a separate PR.

File: Gemfile.lock (hunk 5 above)
Comment (styliii) [reply Gemfile.lock:347]: oh no. i'll change it back then

**Pull request merged**